Watchdog service code refactor

pierantoniomerlino opened the pull request:
Signed-off-by: pierantoniomerlino <pierantonio.merlino@eurotech.com>

MMaiero committed:
Changes to have can tests working on osx.

Signed-off-by: MMaiero <matteo.maiero@eurotech.com>

MMaiero committed:
The following changes were applied:

- Renamed variables to meet the new format.
- Converted thread_done from a class variable to a local variable.
- Extracted terminatePollThread method.

Signed-off-by: MMaiero <matteo.maiero@eurotech.com>

MMaiero committed:
interfaceName, canId and isMaster are now volatile.

Signed-off-by: MMaiero <matteo.maiero@eurotech.com>

pierantoniomerlino committed:
Watchdog service refactor

Signed-off-by: pierantoniomerlino <pierantonio.merlino@eurotech.com>

pierantoniomerlino committed:
Watchdog service refactor

Signed-off-by: pierantoniomerlino <pierantonio.merlino@eurotech.com>

MMaiero commented on kura/org.eclipse.kura.linux.watchdog/src/main/java/org/eclipse/kura/linux/watchdog/CriticalComponentImpl.java:
```diff
@@ -11,24 +11,37 @@
     /**
      *
+     * @param criticalComponent
      * @param name
```
There is a single param here

MMaiero commented on kura/org.eclipse.kura.linux.watchdog/src/main/java/org/eclipse/kura/linux/watchdog/WatchdogServiceImpl.java:
```diff
@@ -11,134 +11,124 @@
     }
 
-    protected void deactivate(ComponentContext componentContext) {
+    protected void deactivate() {
```
I would split this into separate methods in order to segregate the actions taken.

MMaiero committed:
Commented can.test

Signed-off-by: MMaiero <matteo.maiero@eurotech.com>

ctron commented on kura/org.eclipse.kura.linux.watchdog/src/main/java/org/eclipse/kura/linux/watchdog/WatchdogServiceImpl.java:
```diff
@@ -338,6 +319,29 @@ private void readProperties(Map<String, Object> properties) {
+    private void cancelPollThread() {
+        if (pollThreadTask != null && !pollThreadTask.isCancelled()) {
+            logger.debug("Cancelling WatchdogServiceImpl task ...");
+            pollThreadTask.cancel(true);
```
If you shut down the executor and the task was scheduled with this executor, then there is no need to cancel the task itself.

pierantoniomerlino replied:
Yes, I could avoid the task cancellation if I shutdown the executor...

cdealti commented on kura/org.eclipse.kura.linux.watchdog/META-INF/MANIFEST.MF:
```diff
@@ -8,7 +8,10 @@ Bundle-RequiredExecutionEnvironment: JavaSE-1.6
-Import-Package: org.eclipse.kura.configuration;version="[1.1,1.2)",
+Import-Package: org.eclipse.kura;version="[1.2.0,2.0.0)",
+ org.eclipse.kura.configuration;version="[1.1,1.2)",
+ org.eclipse.kura.core.util;version="[1.1.0,2.0.0)",
```
We usually import packages omitting the micro version. It should be [1.1,2.0)

cdealti commented on kura/org.eclipse.kura.linux.watchdog/META-INF/MANIFEST.MF:
```diff
@@ -8,7 +8,10 @@ Bundle-RequiredExecutionEnvironment: JavaSE-1.6
 Bundle-ActivationPolicy: lazy
-Import-Package: org.eclipse.kura.configuration;version="[1.1,1.2)",
+Import-Package: org.eclipse.kura;version="[1.2.0,2.0.0)",
+ org.eclipse.kura.configuration;version="[1.1,1.2)",
```
Here we are implementing org.eclipse.kura.configuration.ConfigurableComponent as consumers.
See [Wiki page](https://github.com/eclipse/kura/wiki/Kura-Semantic-Versioning).
Hence we need to import it with [1.1,2.0)

cdealti commented on kura/org.eclipse.kura.linux.watchdog/META-INF/MANIFEST.MF:
```diff
@@ -8,7 +8,10 @@ Bundle-RequiredExecutionEnvironment: JavaSE-1.6
 Bundle-ClassPath: .
 Bundle-ActivationPolicy: lazy
-Import-Package: org.eclipse.kura.configuration;version="[1.1,1.2)",
+Import-Package: org.eclipse.kura;version="[1.2.0,2.0.0)",
```
Same as below. Omit micro version

cdealti commented on kura/org.eclipse.kura.linux.watchdog/OSGI-INF/metatype/org.eclipse.kura.watchdog.WatchdogService.xml:
```diff
@@ -42,6 +42,17 @@
             default="/dev/watchdog"
             description="Watchdog device."/>
+            
+        <AD id="watchdogAction"
```
We have a chance to use a more readable property ``name``. It can be different from ``id`` and the one shown in the Web UI.

cdealti commented on kura/org.eclipse.kura.linux.watchdog/src/main/java/org/eclipse/kura/linux/watchdog/CriticalComponentImpl.java:
```diff
@@ -11,24 +11,37 @@
     }
 
+    public CriticalComponent getCriticalComponent() {
+        return criticalComponent;
```
this.criticalComponent. Run formatter?

cdealti commented on kura/org.eclipse.kura.linux.watchdog/src/main/java/org/eclipse/kura/linux/watchdog/WatchdogServiceImpl.java:
```diff
@@ -11,134 +11,124 @@
+import org.eclipse.kura.KuraErrorCode;
+import org.eclipse.kura.KuraException;
 import org.eclipse.kura.configuration.ConfigurableComponent;
+import org.eclipse.kura.core.util.ProcessUtil;
```
This creates a dependency on the org.eclipse.kura.core bundle.
Please use Apache Commons Exec or plain Java Runtime for the limited purpose of executing the "reboot" command.

cdealti commented on kura/org.eclipse.kura.linux.watchdog/src/main/java/org/eclipse/kura/linux/watchdog/WatchdogServiceImpl.java:
```diff
@@ -11,134 +11,124 @@
+    private int pingInterval = 2000;
     private String watchdogDevice = "/dev/watchdog";
+    private String watchdogAction = WATCHDOG_ACTION_SYSTEM_REBOOT;
+    private volatile boolean configEnabled = false;   // initialized in properties, if false -> no watchdog
```
initialization not needed

cdealti commented on kura/org.eclipse.kura.linux.watchdog/src/main/java/org/eclipse/kura/linux/watchdog/WatchdogServiceImpl.java:
```diff
@@ -11,134 +11,124 @@
 
-        this.m_pollThreadExecutor = Executors.newSingleThreadScheduledExecutor();
+    protected void activate(ComponentContext componentContext, Map<String, Object> properties) {
+        criticalServiceList = Collections.synchronizedList(new ArrayList<CriticalComponentImpl>());
```
this.criticalServiceList. Run formatter?

cdealti commented on kura/org.eclipse.kura.linux.watchdog/src/main/java/org/eclipse/kura/linux/watchdog/WatchdogServiceImpl.java:
```diff
@@ -11,134 +11,124 @@
-        this.m_pollThreadExecutor = Executors.newSingleThreadScheduledExecutor();
+    protected void activate(ComponentContext componentContext, Map<String, Object> properties) {
+        criticalServiceList = Collections.synchronizedList(new ArrayList<CriticalComponentImpl>());
+        this.isEnabled = false;
```
it's better to set initial values in a single place. If you don't want to do that in a constructor, set them here. This includes removing the initialization of private fields here

cdealti commented on kura/org.eclipse.kura.linux.watchdog/src/main/java/org/eclipse/kura/linux/watchdog/WatchdogServiceImpl.java:
```diff
@@ -11,134 +11,124 @@
-            s_pollThreadTask.cancel(true);
-            s_logger.info("WatchdogServiceImpl task cancelled? = {}", s_pollThreadTask.isDone());
-            s_pollThreadTask = null;
+        if (pollThreadTask != null && !pollThreadTask.isDone()) {
```
Run the formatter

cdealti commented on kura/org.eclipse.kura.linux.watchdog/src/main/java/org/eclipse/kura/linux/watchdog/WatchdogServiceImpl.java:
```diff
@@ -11,134 +11,124 @@
-        }
-        s_criticalServiceList = new ArrayList<CriticalComponentImpl>();
-        this.m_enabled = false;
+    private ScheduledFuture<?> pollThreadTask;
```
pollTask is better

cdealti commented on kura/org.eclipse.kura.linux.watchdog/src/main/java/org/eclipse/kura/linux/watchdog/WatchdogServiceImpl.java:
```diff
@@ -11,134 +11,124 @@
-        s_criticalServiceList = new ArrayList<CriticalComponentImpl>();
-        this.m_enabled = false;
+    private ScheduledFuture<?> pollThreadTask;
+    private ScheduledExecutorService pollThreadExecutor;
```
pollExecutor is better

cdealti commented on kura/org.eclipse.kura.linux.watchdog/src/main/java/org/eclipse/kura/linux/watchdog/WatchdogServiceImpl.java:
```diff
@@ -11,134 +11,124 @@
+            this.pollThreadExecutor.shutdownNow();
             try {
-                this.m_pollThreadExecutor.awaitTermination(THREAD_TERMINATION_TOUT, TimeUnit.SECONDS);
+                this.pollThreadExecutor.awaitTermination(THREAD_TERMINATION_TOUT, TimeUnit.MILLISECONDS);
```
why?

cdealti commented on kura/org.eclipse.kura.linux.watchdog/src/main/java/org/eclipse/kura/linux/watchdog/WatchdogServiceImpl.java:
```diff
@@ -11,134 +11,124 @@
                 @Override
                 public void run() {
                     Thread.currentThread().setName("WatchdogServiceImpl");
                     doWatchdogLoop();
```
Not a loop anymore. doWatchdog is enough

amitjoy committed:
Added missing exception in Driver Javadoc

Signed-off-by: Amit Kumar Mondal <admin@amitinside.com>

amitjoy committed:
Added OPC-UA Driver

Signed-off-by: Amit Kumar Mondal <admin@amitinside.com>

amitjoy committed:
Added OPC-UA Driver Localization Resource

Signed-off-by: Amit Kumar Mondal <admin@amitinside.com>

Code Refactoring and Cleanups in OPC-UA Driver Implementation

Signed-off-by: Amit Kumar Mondal <admin@amitinside.com>

Added OPC-UA Driver and its localization source in distribution

Signed-off-by: Amit Kumar Mondal <admin@amitinside.com>

markoer committed:
added System Service unit tests and moved ITs

updated getPrimaryMacAddress() and added getBiosVersion() for Windows
minor refactoring for better testability

Signed-off-by: Marko Erker <marko.erker@comtrade.com>

pierantoniomerlino commented:
@cdealti could you please take a look on this PR?

Commit:
#1079: Tests for org.eclipse.kura.core.comm

- Prepared unit tests for class CommConnectionImpl from package
  org.eclipse.kura.core.comm
- Fixed the implementation of getBytesAsString method (extra space at
  the end of the result string)
- Ported the org.eclipse.kura.core.comm.test project from "kura" to
  "kura/test" directory.

Signed-off-by: Djuro Drljaca <djuro.drljaca@comtrade.com>

pierantoniomerlino committed:
Removed duplicates from Hotspot list

Signed-off-by: pierantoniomerlino <pierantonio.merlino@eurotech.com>

markoer committed:
updated InstallImpl for running batch files
added InstallImpl unit tests
added unit tests of DeploymentPackageInstallOptions

Signed-off-by: Marko Erker <marko.erker@comtrade.com>

Commit:
#1111: Tests for org.eclipse.kura.core.ssl

- Increased BREE from Java 1.6 to Java 1.8 in bundle
  org.eclipse.kura.core
- Removed checking for Java 1.6 in
  SSLSocketFactoryWrapper.updateSSLParameters() method
- Prepared unit tests for ConnectionSslOptions and
  SSLSocketFactoryWrapper classes.

Signed-off-by: Djuro Drljaca <djuro.drljaca@comtrade.com>

nicolatimeus committed:
OPCUA driver fixes/improvements

Added some missing package imports.
Replaced bundled jars with Milo 0.1.0 official Maven deps.
Improved request timeout handling.
Minor code cleanups.

Signed-off-by: nicolatimeus <nicola.timeus@eurotech.com>

ctron committed:
#1123: Fix reported issues

Signed-off-by: Jens Reimann <jreimann@redhat.com>

amitjoy committed:
Added functionality to allow adding Wire Components after removing one

This allows adding Wire Components of same names to the Wire Graph even
after removing a Wire Component of the same name. This does not require
saving the Wire Graph before adding new Wire Components.

Signed-off-by: Amit Kumar Mondal <admin@amitinside.com>

pierantoniomerlino committed:
Fixed installer/upgrader for watchdog

Signed-off-by: pierantoniomerlino <pierantonio.merlino@eurotech.com>

pierantoniomerlino committed:
Fixed installer/upgrader for watchdog

Signed-off-by: pierantoniomerlino <pierantonio.merlino@eurotech.com>

MMaiero committed:
Changes to prevent exception of wires panel refresh.

It can happen that, with heavy components, the getWiresConfigurationInternal() invocation
at the end of updateWireConfiguration() determines an internal error, probably due to a refresh
when the component is not effectively up and running.
The changes try to prevent this refresh, as not really needed.
Small change in WireComponentTrackerCustomizer.java to try to prevent a ConcurrentModificationException.

Signed-off-by: MMaiero <matteo.maiero@eurotech.com>

markoer committed:
relaxed the test a bit to account for slower hosts

Signed-off-by: Marko Erker <marko.erker@comtrade.com>

markoer committed:
Fixed responses and added tests

fixed a few responses in case of exception
added IOBException check
refactored for testability
added tests covering corner cases and exception flows

Signed-off-by: Marko Erker <marko.erker@comtrade.com>

MMaiero committed:
Splitted com.codeminders.hidapi and org.eclipse.soda.dk.comm.

Native parts are now in separate fragments.

Signed-off-by: MMaiero <matteo.maiero@eurotech.com>

MMaiero committed:
Applied formatting to pom files.

Minor bug fixes in fragments Manifest.MF files.

Signed-off-by: MMaiero <matteo.maiero@eurotech.com>

markoer committed:
fixed wrong cast
added comments
updated for testability
added tests of untested code

Signed-off-by: Marko Erker <marko.erker@comtrade.com>

ctron committed:
Implement Kapua's applicationFramework/applicationFrameworkVersion

Kapua actually uses two fields for the gateway version called
"application framework" and "application framework version", compared
to the OSGi framework.

This change does implement this information but still allows backwards
compatibility by providing the "kura_version" information.

Signed-off-by: Jens Reimann <jreimann@redhat.com>

apupier committed:
Provide better Separation of Concerns

- extract classes to have a more readable Camel route

Signed-off-by: Aurelien Pupier <apupier@redhat.com>

MMaiero committed:
Merge pull request #1086 from amitjoy/PR_opcua_driver

OPC-UA Driver

MMaiero committed:
Merge pull request #1130 from markoer/ut_cfg_svc_relax

Relaxed a test a bit to account for slower hosts

MMaiero committed:
Solved provider.test failures.

Signed-off-by: MMaiero <matteo.maiero@eurotech.com>

MMaiero committed:
Version changes.

Increased org.eclipse.kura.core.net imported version of wifi api to allow compiling.
Changes in org.eclipse.kura.linux.net.test, org.eclipse.kura.net.admin and
org.eclipse.kura.web2 for compatibility purposes.

Signed-off-by: MMaiero <matteo.maiero@eurotech.com>

pierantoniomerlino committed:
Merge pull request #1135 from eclipse/split_native-libs

Splitted com.codeminders.hidapi and org.eclipse.soda.dk.comm.

MMaiero committed:
Merge pull request #1087 from markoer/utit_system_service

System Service updates and unit tests

MMaiero committed:
Merge pull request #1141 from nicolatimeus/fix_opcua-driver

OPCUA driver fixes/improvements

MMaiero committed:
Merge pull request #1080 from djurod/PR-Tests-core.comm.CommConnectionImpl

#1079: Tests for org.eclipse.kura.core.comm

MMaiero committed:
Merge pull request #1110 from eclipse/fix_getWifiHotspots

Fix for SSID search

MMaiero committed:
Merge pull request #1124 from ctron/feature/fix_issue_1123

#1123: Fix reported issues

nicolatimeus committed:
Added support for enabling serial rx timeout

Signed-off-by: nicolatimeus <nicola.timeus@eurotech.com>

markoer committed:
added 2 plugins and worspaces

Signed-off-by: Marko Erker <marko.erker@comtrade.com>

MMaiero committed:
Merge pull request #1143 from nicolatimeus/enh_serial-rx-timeout

Added support for enabling serial rx timeout

MMaiero committed:
Merge pull request #1045 from eclipse/fix_can.test.mac

Changes to have can tests working.

MMaiero committed:
Merge pull request #1142 from eclipse/fix_wire-provider-test

Solved provider.test failures.

MMaiero committed:
Merge branch 'PR-Tests-core.ssl' of https://github.com/djurod/kura into djurod-PR-Tests-core.ssl

MMaiero committed:
Merge pull request #1129 from markoer/utit_deployment_install

Added InstallImpl and DeploymentPackageInstallOptions unit tests

MMaiero committed:
Updated copyrights.

Signed-off-by: MMaiero <matteo.maiero@eurotech.com>

MMaiero committed:
Merge branch 'djurod-PR-Tests-core.ssl' into develop

MMaiero committed:
Merge pull request #1127 from eclipse/fix_wires-refresh-error

Changes to prevent exception of wires panel refresh.

MMaiero committed:
Merge branch 'feature/app_fw_version_1' of https://github.com/ctron/kura into ctron-feature/app_fw_version_1

MMaiero committed:
Merge branch 'ctron-feature/app_fw_version_1' into develop

markoer committed:
fixed project versions, added javadoc dependency

Signed-off-by: Marko Erker <marko.erker@comtrade.com>

ctron committed:
Implement new Cloud Client API (#1122)

This change implements methods which previously threw "unsupported operation" exceptions.

It also cleans up a few minor things:

 * Add error code SUBSCRIPTION_ERROR
 * Add the ability to control JMX support
 * Increment versions
 * Minor cleanups, switch to Java 8
 * Add some javadoc

Signed-off-by: Jens Reimann <jreimann@redhat.com>

ctron committed:
Add missing Kura wires bundles, breaks the Web UI

MMaiero committed:
Merge pull request #1145 from markoer/pr_project_structure_fix

Fixed project versions, added javadoc dependency

ctron committed:
Enable JMX by default, allow overriding using system properties

Signed-off-by: Jens Reimann <jreimann@redhat.com>

ctron committed:
Switch over to Java 8 and use the functional interface

Signed-off-by: Jens Reimann <jreimann@redhat.com>

nicolatimeus committed:
Implemented alphabetic sorting of dropdown lists in the Web Ui.

Signed-off-by: nicolatimeus <nicola.timeus@eurotech.com>

nicolatimeus committed:
Use window.parent instead of top window for interface between wires GWT and js code.

Signed-off-by: nicolatimeus <nicola.timeus@eurotech.com>

MMaiero committed:
Modified web ui activation message at startup.

Minor changes to meet the new code styling.

Signed-off-by: MMaiero <matteo.maiero@eurotech.com>

MMaiero committed:
Merge pull request #1149 from ctron/feature/cleanup_camel_1

Switch over to Java 8 and use the functional interface

MMaiero committed:
Merge pull request #1154 from nicolatimeus/enh_web-ui-self-contained

Use window.parent instead of top in wires js and GWT parts

MMaiero committed:
Merge pull request #1146 from markoer/utit_deployment_handler

[Testing] CloudDeploymentHandlerV2: Fixed responses and added tests

MMaiero committed:
Merge pull request #1147 from markoer/ut_deployment_download

[Test&Fix] DownloadImpl: refactored and added tests

nicolatimeus committed:
Improved name check when creating new wire components

Signed-off-by: nicolatimeus <nicola.timeus@eurotech.com>

amitjoy committed:
Fixed Wrong Order of Operation Execution

Signed-off-by: Amit Kumar Mondal <admin@amitinside.com>

pierantoniomerlino committed:
Modified Getting Started section on README.md

Signed-off-by: pierantoniomerlino <pierantonio.merlino@eurotech.com>

MMaiero committed:
Merge pull request #1167 from amitjoy/PR_order_fix

Fixed Wrong Order of Operation Execution

MMaiero committed:
Prevented editability of fixed fields.

Changed the server side of GWT code in order to ignore a malicious change to a
fixed field when a component configuration is updated.

Signed-off-by: MMaiero <matteo.maiero@eurotech.com>

MMaiero committed:
Merge pull request #1126 from amitjoy/PR_savegraph

Added functionality to allow adding Wire Components after removing one

amitjoy committed:
OPC-UA Driver: Minor Fixes and Cleanup

Signed-off-by: Amit Kumar Mondal <admin@amitinside.com>

ctron committed:
Apply Kura formatter and cleanup profile

Signed-off-by: Jens Reimann <jreimann@redhat.com>

MMaiero committed:
Merge pull request #1173 from amitjoy/PR_minor_opc_fixes

OPC-UA Driver: Minor Fixes and Cleanup

MMaiero committed:
Merge pull request #1153 from nicolatimeus/enh_config-sort-dropdowns

Implemented alphabetic sorting of dropdown lists in the Web Ui.

MMaiero committed:
Merge pull request #1170 from nicolatimeus/fix_wires-invalid-component-names

Improved name check when creating new wire components

ctron committed:
Minor cleanups for the camel aggregator example

ctron committed:
Change version to 1.1.1-SNAPSHOT for the camel aggregation example

ctron committed:
Switch default behavior to only track relevant services

Also change private fields to new Kura style.

Signed-off-by: Jens Reimann <jreimann@redhat.com>

MMaiero committed:
Uniformed save/delete dialogs in wires with what in the rest of the ui.

Signed-off-by: MMaiero <matteo.maiero@eurotech.com>

markoer committed:
added wires unit and integration tests

Signed-off-by: Marko Erker <marko.erker@comtrade.com>

amitjoy committed:
Fix Javadoc and Method Name in DbServiceHelper for better comprehension

Signed-off-by: Amit Kumar Mondal <admin@amitinside.com>

ctron committed:
Drop support for "fedorapi"

Signed-off-by: Jens Reimann <jreimann@redhat.com>

ctron committed:
Adding a "non-network" version for Fedora 25

Signed-off-by: Jens Reimann <jreimann@redhat.com>

MMaiero committed:
Merge pull request #1179 from ctron/feature/drop_fedorapi_1

Drop support for "fedorapi"

MMaiero committed:
Merge pull request #1171 from eclipse/improve_readme

Modified Getting Started section on README.md

MMaiero committed:
Merge pull request #1177 from amitjoy/PR_fix_doc

Fix Javadoc and Method Name in DbServiceHelper for better comprehension

MMaiero committed:
Merge pull request #1178 from markoer/it_wire_provider_test

[Testing] Added Wires tests

MMaiero committed:
Implemented changes to prevent eth0 toggling when wlan0 is disabled.

Signed-off-by: MMaiero <matteo.maiero@eurotech.com>

cdealti committed:
Partially revert net.wifi breaking change

This commit partially revert an API breaking change introduced by 681c285
and specifically part of the refactor of o.e.k.net.wifi.WifiHotspotInfo.
Replacing EnumSet with Set in return types is not backward compatible.

Signed-off-by: Cristiano De Alti <cristiano.dealti@eurotech.com>

MMaiero committed:
Merge branch 'feature/fedora25-nn-1' of https://github.com/ctron/kura into ctron-feature/fedora25-nn-1

Signed-off-by: MMaiero <matteo.maiero@eurotech.com>

MMaiero committed:
Modifications to fedora25-nn profile.

Removed unneeded files, cleanups to have it as a proper nn profile.

Signed-off-by: MMaiero <matteo.maiero@eurotech.com>

MMaiero committed:
Merge branch 'ctron-feature/fedora25-nn-1' into develop

ctron committed:
Add feature for Camel API

Signed-off-by: Jens Reimann <jreimann@redhat.com>

MMaiero committed:
Modified the cleanup query.

Signed-off-by: MMaiero <matteo.maiero@eurotech.com>

MMaiero committed:
Merge pull request #1175 from eclipse/fix_wires-save-appearance

Uniformed save/delete dialogs in wires with what in the rest of the ui.

MMaiero committed:
Merge pull request #1184 from cdealti/pr/revert-net-wifi-breaking-change

Partially revert net.wifi breaking change

MMaiero committed:
Merge pull request #1172 from eclipse/fix_editable-fixed-fields

Prevented editability of fixed fields.

MMaiero committed:
Merge pull request #1157 from eclipse/fix_systemServiceMessage

Modified web ui activation message at startup.

Commit:
#1163: Tests for org.eclipse.kura.wire.component.provider

- Prepared unit tests for WireAsset, DbWireRecordFilter, EmitJob and
  Timer classes.
- Changed Timer class so that it is possible to unit test it.

Signed-off-by: Djuro Drljaca <djuro.drljaca@comtrade.com>

MMaiero committed:
Fixed long ranges and property description.

Signed-off-by: MMaiero <matteo.maiero@eurotech.com>

cdealti committed:
WireEmitter/Receiver are ConsumerTypeS

The only way an application can consume these types is to
implement them.
An real-world graph will always include domain specific
WireComponentS (WireEmitterS and WireReceiverS) that
must developed by implementing and registering these
interfaces.

Signed-off-by: Cristiano De Alti <cristiano.dealti@eurotech.com>

MMaiero committed:
Preventing possible NPE when MetaTypeInformation is null.

Signed-off-by: MMaiero <matteo.maiero@eurotech.com>

cdealti committed:
Fix iptables-restore run in background

In the firewall init script iptables-restore is run in background.
This is not usually a problem unless the firewall_cust script
exists.
In this case the firewall init script will create iptable
rules concurrently often leading to a state where some
rules are missing.

Signed-off-by: Cristiano De Alti <cristiano.dealti@eurotech.com>

markoer committed:
add directory if it doesn't exist

Signed-off-by: Marko Erker <marko.erker@comtrade.com>

MMaiero committed:
Merge pull request #1187 from eclipse/fix_dbwirecomp-cleanup

Modified the cleanup query.

MMaiero committed:
Changed wires component selection.

With this change in place it should be possible to select a component and have
the configuration shown without the need to click first outside the component
in the composer.

Signed-off-by: MMaiero <matteo.maiero@eurotech.com>

Commit:
Merge pull request #1156 from eclipse/fix_eth-disable-wlan

Implemented changes to prevent eth0 toggling when wlan0 is disabled.

ctron committed:
Mark Karaf plugins for M2E as ignored

MMaiero committed:
Merge pull request #1198 from eclipse/fix_wires-comp-selection

Changed wires component selection.

MMaiero committed:
Merge pull request #1192 from cdealti/pr/are_wire_emitter_receiver_consumer_types

WireEmitter/Receiver are ConsumerTypeS

MMaiero committed:
Merge pull request #1183 from djurod/PR-Tests-wire.component.provider

#1163: Tests for org.eclipse.kura.wire.component.provider

nicolatimeus committed:
Modified Asset and Driver APIs to add support for prepared reads.

Signed-off-by: nicolatimeus <nicola.timeus@eurotech.com>

MMaiero committed:
Merge pull request #1190 from eclipse/fix_example-pub-long-ranges

Fixed long ranges and property description.

MMaiero committed:
Merge pull request #1199 from markoer/pr_cfg_svc_test_mkdir

[Testing] Add snapshot directory if it doesn't exist

MMaiero committed:
Merge pull request #1194 from cdealti/pr/fix_firewall_init_script

Fix iptables-restore run in background

MMaiero committed:
Merge pull request #1191 from eclipse/fix_MetaTypeInformation-npe

Preventing possible NPE when MetaTypeInformation is null.

MMaiero committed:
Modified ui message for values out of range.

Signed-off-by: MMaiero <matteo.maiero@eurotech.com>

MMaiero committed:
Merge pull request #1189 from eclipse/fix_out-range-param-message

Modified ui message for values out of range.

MMaiero committed:
Merge pull request #1174 from ctron/feature/default_track_only_1

Switch default behavior to only track relevant services

nicolatimeus committed:
Installer script fixes

Signed-off-by: nicolatimeus <nicola.timeus@eurotech.com>

MMaiero committed:
Fixed bundles versions.

Changed bundles versions in order to get, in distrib, the all the bundles that are
configurable register this in the framework, in order to be tracked by
configuration service and web ui.

Signed-off-by: MMaiero <matteo.maiero@eurotech.com>

ctron committed:
Make it a functional interface

Signed-off-by: Jens Reimann <jreimann@redhat.com>

MMaiero committed:
Merge pull request #1200 from ctron/feature/plain_maven_1

Add feature for Camel API

MMaiero committed:
Merge pull request #1224 from ctron/feature/java_8_camel_1

Make it a functional interface

ctron committed:
Allow providing an input stream for reading XML content

Signed-off-by: Jens Reimann <jreimann@redhat.com>

ctron committed:
Relax null handling, fix possible NPE, improve javadoc

Signed-off-by: Jens Reimann <jreimann@redhat.com>

ctron committed:
Add a note about the Fedora firewall

Signed-off-by: Jens Reimann <jreimann@redhat.com>

amitjoy committed:
Fixed Asset Persistency in Snapshot and Refactoring

Fixes #1222

Signed-off-by: Amit Kumar Mondal <admin@amitinside.com>

MMaiero committed:
Merge pull request #1232 from amitjoy/PR_fix_asset_persistency

Fixed Asset Persistency in Snapshot and Refactoring

MMaiero committed:
Merge pull request #1227 from ctron/feature/fix_possible_npe_1

Relax null handling, fix possible NPE, improve javadoc

MMaiero committed:
Merge pull request #1226 from ctron/feature/xml_camel_ctor_1

Allow providing an input stream for reading XML content

MMaiero committed:
Merge pull request #1221 from eclipse/fix_bundles-versions

Fixed bundles versions.

MMaiero committed:
Merge pull request #1215 from nicolatimeus/fix_installer

Installer script fixes

MMaiero committed:
Removed timeout based on magic number.

Introduced a proper service tracker. This allows a proper component and ui management
with all the types of devices.

Signed-off-by: MMaiero <matteo.maiero@eurotech.com>

amitjoy committed:
Merge pull request #1182 from eclipse/fix_magic-number-timeout

Removed timeout based on magic number.

pierantoniomerlino committed:
Fixed gpsDevice object

Signed-off-by: pierantoniomerlino <pierantonio.merlino@eurotech.com>

ctron committed:
Stick to Equinox Neon.1 from archived sources

Signed-off-by: Jens Reimann <jreimann@redhat.com>

MMaiero committed:
Merge pull request #1235 from ctron/feature/use_archived_neon1_1

Stick to Equinox Neon.1 from archived sources

MMaiero committed:
Merge pull request #1233 from pierantoniomerlino/fix_gps

Fixed gpsDevice object

cdealti commented on kura/org.eclipse.kura.linux.watchdog/META-INF/MANIFEST.MF:
```diff
@@ -8,7 +8,10 @@ Bundle-RequiredExecutionEnvironment: JavaSE-1.6
+Import-Package: org.apache.commons.exec;version="1.3.0",
+ org.eclipse.kura;version="[1.2,2.0)",
+ org.eclipse.kura.configuration;version="[1.1,2.0)",
+ org.eclipse.kura.watchdog;version="[1.0,2.0)",
```
``org.eclipse.kura.watchdog;version="[1.0,1.1)"`` since this bundle **provides** the implementation

cdealti commented on kura/org.eclipse.kura.linux.watchdog/META-INF/MANIFEST.MF:
```diff
@@ -8,7 +8,10 @@ Bundle-RequiredExecutionEnvironment: JavaSE-1.6
-Import-Package: org.eclipse.kura.configuration;version="[1.1,1.2)",
- org.eclipse.kura.watchdog;version="[1.0,1.1)",
+Import-Package: org.apache.commons.exec;version="1.3.0",
+ org.eclipse.kura;version="[1.2,2.0)",
```
``org.eclipse.kura;version="[1.3,2.0)",`` since this is the version on develop

cdealti commented on kura/org.eclipse.kura.linux.watchdog/OSGI-INF/metatype/org.eclipse.kura.watchdog.WatchdogService.xml:
```diff
@@ -20,29 +20,29 @@
         
         <AD id="enabled"  
-            name="enabled"
+            name="Enable the watchdog"
```
"Watchdog enable"

cdealti replied:
Please also change the description of the OCD element to:
```
The WatchdogService monitors CriticalComponentS and reboots the system if one of them hangs.
Once enabled the WatchdogService starts refreshing the hardware watchdog which will
reset the system if WatchdogService hangs.
```

MMaiero committed:
Added Neon.1 mirror.

Signed-off-by: MMaiero <matteo.maiero@eurotech.com>

cdealti commented on kura/org.eclipse.kura.linux.watchdog/META-INF/MANIFEST.MF:
```diff
@@ -8,7 +8,10 @@ Bundle-RequiredExecutionEnvironment: JavaSE-1.6
 Bundle-ActivationPolicy: lazy
-Import-Package: org.eclipse.kura.configuration;version="[1.1,1.2)",
- org.eclipse.kura.watchdog;version="[1.0,1.1)",
+Import-Package: org.apache.commons.exec;version="1.3.0",
```
Unused

cdealti commented on kura/org.eclipse.kura.linux.watchdog/META-INF/MANIFEST.MF:
```diff
@@ -8,7 +8,10 @@ Bundle-RequiredExecutionEnvironment: JavaSE-1.6
+ org.eclipse.kura;version="[1.2,2.0)",
+ org.eclipse.kura.configuration;version="[1.1,2.0)",
+ org.eclipse.kura.watchdog;version="[1.0,2.0)",
+ org.osgi.framework;version="1.5.0",
```
org.osgi.framework and org.osgi.service.component are unused

MMaiero committed:
Merge pull request #1237 from eclipse/enh_add-equinox-mirror

Added Neon.1 mirror.

MMaiero committed:
Merge pull request #1225 from nicolatimeus/enh_wires-prepared-read

First implementation of prepared reads for Asset and Driver

nicolatimeus committed:
Changed return type of Driver.read() and Driver.write() to void

Signed-off-by: nicolatimeus <nicola.timeus@eurotech.com>

cdealti commented on kura/org.eclipse.kura.linux.watchdog/OSGI-INF/metatype/org.eclipse.kura.watchdog.WatchdogService.xml:
```diff
@@ -20,29 +20,29 @@
         
         <AD id="pingInterval"
-            name="pingInterval"
+            name="Watchdog refresh interval"
```
I'd change the description in "Watchdog refresh interval in ms...". I've never understood the word "ping"

cdealti replied:
Can we add a max value of 60000 ms seconds? Accordingly to the kernel documentation this is the max interval.

cdealti commented on kura/org.eclipse.kura.linux.watchdog/OSGI-INF/metatype/org.eclipse.kura.watchdog.WatchdogService.xml:
```diff
@@ -20,29 +20,29 @@
 
         <AD id="watchdogDevice"
-            name="watchdogDevice"
+            name="Watchdog device path"
```
Change the description to "Watchdog device path e.g. /dev/watchdog"

cdealti commented on kura/org.eclipse.kura.linux.watchdog/src/main/java/org/eclipse/kura/linux/watchdog/CriticalComponentImpl.java:
```diff
@@ -11,24 +11,34 @@
-     * @param name
-     * @param timeout
-     *            timeout for reporting interval in seconds
+     * @param criticalComponent
```
I think we can remove the Javadoc unless Sonar Lint complains

cdealti commented on kura/org.eclipse.kura.linux.watchdog/src/main/java/org/eclipse/kura/linux/watchdog/CriticalComponentImpl.java:
```diff
@@ -11,24 +11,34 @@
+
 public class CriticalComponentImpl {
 
+    private CriticalComponent criticalComponent;
```
More than an CriticalComponentImpl this is an CriticalComponentImpl element or entry in the list of registered CriticalComponents

cdealti commented on kura/org.eclipse.kura.linux.watchdog/src/main/java/org/eclipse/kura/linux/watchdog/CriticalComponentImpl.java:
```diff
@@ -11,24 +11,34 @@
+        return criticalComponent;
+    }
+
+    public void setCriticalComponent(CriticalComponent criticalComponent) {
```
Remove setter

cdealti commented on kura/org.eclipse.kura.linux.watchdog/src/main/java/org/eclipse/kura/linux/watchdog/WatchdogServiceImpl.java:
```diff
@@ -14,131 +14,75 @@
+    private int pingInterval;
+    private String watchdogDevice;
+    private volatile boolean configEnabled;   // initialized in properties, if false -> no watchdog
+    private List<CriticalComponentImpl> criticalServiceList;
```
criticalServiceList -> criticalComponentList

cdealti commented on kura/org.eclipse.kura.linux.watchdog/src/main/java/org/eclipse/kura/linux/watchdog/WatchdogServiceImpl.java:
```diff
@@ -14,131 +14,75 @@
+    private volatile boolean watchdogToStop;
+    private int pingInterval;
+    private String watchdogDevice;
+    private volatile boolean configEnabled;   // initialized in properties, if false -> no watchdog
```
initialized in properties -> enabled in configuration

cdealti commented on kura/org.eclipse.kura.linux.watchdog/src/main/java/org/eclipse/kura/linux/watchdog/WatchdogServiceImpl.java:
```diff
@@ -14,131 +14,75 @@
 
-    private static ScheduledFuture<?> s_pollThreadTask;
-    private ScheduledExecutorService m_pollThreadExecutor;
+    private volatile boolean isEnabled;
```
I believe "enabled" is just enough

cdealti commented on kura/org.eclipse.kura.linux.watchdog/src/main/java/org/eclipse/kura/linux/watchdog/WatchdogServiceImpl.java:
```diff
@@ -159,23 +103,19 @@ public int getHardwareTimeout() {
+        boolean existing = false;
+        for (CriticalComponentImpl csi : this.criticalServiceList) {
+            if (criticalComponent.equals(csi.getCriticalComponent())) {
+                existing = true;
```
add break statement

cdealti commented on kura/org.eclipse.kura.linux.watchdog/src/main/java/org/eclipse/kura/linux/watchdog/WatchdogServiceImpl.java:
```diff
@@ -159,23 +103,19 @@ public int getHardwareTimeout() {
 
-        s_logger.debug("Added " + criticalComponent.getCriticalComponentName() + ", with timeout = "
-                + criticalComponent.getCriticalComponentTimeout() + ", list contains " + s_criticalServiceList.size()
+        logger.debug("Added " + criticalComponent.getCriticalComponentName() + ", with timeout = "
```
Use a StringBuilder

cdealti commented on kura/org.eclipse.kura.linux.watchdog/src/main/java/org/eclipse/kura/linux/watchdog/WatchdogServiceImpl.java:
```diff
@@ -159,23 +103,19 @@ public int getHardwareTimeout() {
-        s_logger.debug("Added " + criticalComponent.getCriticalComponentName() + ", with timeout = "
-                + criticalComponent.getCriticalComponentTimeout() + ", list contains " + s_criticalServiceList.size()
+        logger.debug("Added " + criticalComponent.getCriticalComponentName() + ", with timeout = "
+                + criticalComponent.getCriticalComponentTimeout() + ", list contains " + this.criticalServiceList.size()
```
critical services -> critical components

cdealti commented on kura/org.eclipse.kura.linux.watchdog/src/main/java/org/eclipse/kura/linux/watchdog/WatchdogServiceImpl.java:
```diff
@@ -187,13 +127,11 @@ public void registerCriticalService(CriticalComponent criticalComponent) {
+        for (CriticalComponentImpl csi : this.criticalServiceList) {
+            if (criticalComponent.equals(csi.getCriticalComponent())) {
+                this.criticalServiceList.remove(csi);
+                logger.debug("Critical service " + criticalComponent.getCriticalComponentName() + " removed, "
```
same as above

cdealti commented on kura/org.eclipse.kura.linux.watchdog/src/main/java/org/eclipse/kura/linux/watchdog/WatchdogServiceImpl.java:
```diff
@@ -187,13 +127,11 @@ public void registerCriticalService(CriticalComponent criticalComponent) {
+            if (criticalComponent.equals(csi.getCriticalComponent())) {
+                this.criticalServiceList.remove(csi);
+                logger.debug("Critical service " + criticalComponent.getCriticalComponentName() + " removed, "
+                        + System.currentTimeMillis());
```
Add break statement

cdealti commented on kura/org.eclipse.kura.linux.watchdog/src/main/java/org/eclipse/kura/linux/watchdog/WatchdogServiceImpl.java:
```diff
@@ -206,118 +144,149 @@ public void unregisterCriticalService(CriticalComponent criticalComponent) {
     @Override
     public List<CriticalComponent> getCriticalComponents() {
-        return null;
+        List<CriticalComponent> componentList = new ArrayList<CriticalComponent>();
```
return an unmodifiable view with Collections.unmodifiableList

cdealti commented on kura/org.eclipse.kura.linux.watchdog/src/main/java/org/eclipse/kura/linux/watchdog/WatchdogServiceImpl.java:
```diff
@@ -206,118 +144,149 @@ public void unregisterCriticalService(CriticalComponent criticalComponent) {
-                }
+        for (CriticalComponentImpl csi : this.criticalServiceList) {
+            if (criticalService.equals(csi.getCriticalComponent())) {
+                csi.update();
```
add break statement

cdealti commented on kura/org.eclipse.kura.linux.watchdog/src/main/java/org/eclipse/kura/linux/watchdog/WatchdogServiceImpl.java:
```diff
@@ -206,118 +144,149 @@ public void unregisterCriticalService(CriticalComponent criticalComponent) {
+            for (CriticalComponentImpl csi : this.criticalServiceList) {
+                if (csi.isTimedOut()) {
+                    failure = true;
+                    logger.warn("Critical service {} failed -> SYSTEM REBOOT", csi.getName());
```
Critical service -> Critical component

cdealti commented on kura/org.eclipse.kura.linux.watchdog/src/main/java/org/eclipse/kura/linux/watchdog/WatchdogServiceImpl.java:
```diff
@@ -206,118 +144,149 @@ public void unregisterCriticalService(CriticalComponent criticalComponent) {
+    private synchronized void writeWatchdogDevice(String value) {
+        File f = null;
+        FileWriter bw = null;
+        try {
```
we can use try-with-resources statement and avoid the finally clause

cdealti commented on kura/org.eclipse.kura.linux.watchdog/src/main/java/org/eclipse/kura/linux/watchdog/WatchdogServiceImpl.java:
```diff
@@ -206,118 +144,149 @@ public void unregisterCriticalService(CriticalComponent criticalComponent) {
+        // isEnabled is set if the hw watchdog device is activated
+        if (this.isEnabled) {
+            boolean failure = false;
+            // Critical Services
```
Critical Services -> Critical Components

cdealti commented on kura/org.eclipse.kura.linux.watchdog/src/main/java/org/eclipse/kura/linux/watchdog/WatchdogServiceImpl.java:
```diff
@@ -206,118 +144,149 @@ public void unregisterCriticalService(CriticalComponent criticalComponent) {
-                    }
+                    performWatchdogActions();
+                } catch (KuraException e) {
+                    logger.error("Failed to perform watchdog actions.", e);
```
What should happen if this fails? For example if performWatchdogActions fails to reboot the system?
I'd put the body of performWatchdogActions inline. It should be easier to follow.

pierantoniomerlino replied:
The hw watchdog will reset the device in any case.

cdealti commented on kura/org.eclipse.kura.linux.watchdog/src/main/java/org/eclipse/kura/linux/watchdog/WatchdogServiceImpl.java:
```diff
@@ -206,118 +144,149 @@ public void unregisterCriticalService(CriticalComponent criticalComponent) {
+        try {
+            Process proc = Runtime.getRuntime().exec(command);
+            proc.waitFor();
+        } catch (IOException e) {
```
you can catch multiple Exceptions in a single catch block

cdealti commented on kura/org.eclipse.kura.linux.watchdog/src/main/java/org/eclipse/kura/linux/watchdog/WatchdogServiceImpl.java:
```diff
@@ -206,118 +144,149 @@ public void unregisterCriticalService(CriticalComponent criticalComponent) {
-                    if (bw != null) {
-                        bw.close();
-                    }
+                    wdWriter = new PrintWriter(WD_FILE);
```
try-with-resources

cdealti commented on kura/org.eclipse.kura.linux.watchdog/src/main/java/org/eclipse/kura/linux/watchdog/WatchdogServiceImpl.java:
```diff
@@ -206,118 +144,149 @@ public void unregisterCriticalService(CriticalComponent criticalComponent) {
+    }
+
+    private void readProperties(Map<String, Object> properties) {
+        if (properties.get(PROPERTY_PING_INTERVAL) != null) {
```
If you don't rely on the properties also initialize explicitly fields to safe values (same as Metatype defaults IMHO)

cdealti commented on kura/org.eclipse.kura.linux.watchdog/src/main/java/org/eclipse/kura/linux/watchdog/WatchdogServiceImpl.java:
```diff
@@ -206,118 +144,149 @@ public void unregisterCriticalService(CriticalComponent criticalComponent) {
-            FileWriter bw = null;
+        if (this.isEnabled) {
+            writeWatchdogDevice("V");
+            this.isEnabled = false;
```
shouldn't we remove the WD_FILE?

pierantoniomerlino replied:
It's used by the installer/upgrader to try to understand if the watchdog was enabled and refresh it if needed.

cdealti commented on kura/org.eclipse.kura.linux.watchdog/src/main/java/org/eclipse/kura/linux/watchdog/WatchdogServiceImpl.java:
```diff
@@ -206,118 +144,149 @@ public void unregisterCriticalService(CriticalComponent criticalComponent) {
-                    }
+    private void checkCriticalComponents() {
+        // isEnabled is set if the hw watchdog device is activated
+        if (this.isEnabled) {
```
Ideally this should be called **if and only if** the service is enabled in config. If it detects any CriticalComponent which has timed out it must reboot the system. In every case it will refresh the hardware watchdog.
I'd try to remove the isEnabled and watchdogToStop flags entirely because it's difficult to understand where they are set and what they really mean.

pierantoniomerlino replied:
The meaning of isEnabled and watchdogToStop are, indeed, not so clear.
isEnabled means that the hw watchdog has already been enabled and it is used to avoid that a CriticalComponent, already timed out at first run of the checkCriticalComponent, never kicks refreshWatchdog. It was already present before the refactor and I left it unchanged.
watchdogToStop, instead, is used to avoid the multiple execution of the "reboot" command.
They indeed manage corner cases and they can be replaced or removed completely.

cdealti commented on kura/distrib/src/main/deb/control/postinst:
```diff
@@ -3,10 +3,50 @@
+    if [ -f "${WD_TMP_FILE}" ]; then
+        WATCHDOG_DEVICE=`cat ${WD_TMP_FILE}`
+        echo "Got watchdog ${WATCHDOG_DEVICE}" >> ${LOG} 2>&1
+        refreshWatchdog &
```
I'm paranoid but **unconditionally** refreshing the watchdog from a child process is generally not considered a good practice. The parent could die or hang (for whatever reason) and the child process will happily continue to kick the watchdog. We need to find a better way

amitjoy committed:
Merge pull request #1247 from nicolatimeus/enh_driver-void-return-value

Changed return type of Driver.read() and Driver.write() to void

pierantoniomerlino committed:
Removed sysctl.kura.raspbian.conf

Signed-off-by: pierantoniomerlino <pierantonio.merlino@eurotech.com>

pierantoniomerlino committed:
Version fix and code improvements

Signed-off-by: pierantoniomerlino <pierantonio.merlino@eurotech.com>

pierantoniomerlino committed:
Modifed installation scripts

Signed-off-by: pierantoniomerlino <pierantonio.merlino@eurotech.com>

pierantoniomerlino committed:
Fixes

Signed-off-by: pierantoniomerlino <pierantonio.merlino@eurotech.com>

pierantoniomerlino committed:
Fixes Again

Signed-off-by: pierantoniomerlino <pierantonio.merlino@eurotech.com>

pierantoniomerlino committed:
Modified extract.sh script

Signed-off-by: pierantoniomerlino <pierantonio.merlino@eurotech.com>

pierantoniomerlino commented:
Probably something went wrong on this branch after rebasing...

pierantoniomerlino commented:
Closed this and reopened a new one.
